source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,538 @@
+
+// returns true if we should execute an iteration, false otherwise
+static bool begin_iteration(struct scheduled_service_entry *entry) {
+    bool should_execute_iteration = false;
+    struct dispatch_loop_context *contxt = entry->dispatch_queue_context;
+    aws_mutex_lock(&contxt->lock);
+
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+    if (!dispatch_loop) {
+        aws_mutex_unlock(&contxt->lock);
+        return should_execute_iteration;
+    }
+
+    // swap the cross-thread tasks into task-local data
+    AWS_FATAL_ASSERT(aws_linked_list_empty(&dispatch_loop->local_cross_thread_tasks));
+    aws_linked_list_swap_contents(
+        &dispatch_loop->synced_task_data.cross_thread_tasks, &dispatch_loop->local_cross_thread_tasks);
+
+    // mark us as running an iteration and remove from the pending list
+    dispatch_loop->synced_task_data.context->scheduling_state.will_schedule = true;
+    aws_linked_list_remove(&entry->node);
+    aws_mutex_unlock(&contxt->lock);
+
+    should_execute_iteration = true;
+    return should_execute_iteration;
+}
+
+// conditionally schedule another iteration as needed
+static void end_iteration(struct scheduled_service_entry *entry) {
+
+    struct dispatch_loop_context *contxt = entry->dispatch_queue_context;
+    aws_mutex_lock(&contxt->lock);
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+    if (!dispatch_loop) {
+        aws_mutex_unlock(&contxt->lock);
+        return;
+    }
+
+    dispatch_loop->synced_task_data.context->scheduling_state.will_schedule = false;
+
+    // if there are any cross-thread tasks, reschedule an iteration for now
+    if (!aws_linked_list_empty(&dispatch_loop->synced_task_data.cross_thread_tasks)) {
+        // added during service which means nothing was scheduled because will_schedule was true
+        s_try_schedule_new_iteration(contxt, 0);
+    } else {
+        // no cross thread tasks, so check internal time-based scheduler
+        uint64_t next_task_time = 0;
+        /* we already know it has tasks, we just scheduled one. We just want the next run time. */
+        bool has_task = aws_task_scheduler_has_tasks(&dispatch_loop->scheduler, &next_task_time);
+
+        if (has_task) {
+            // only schedule an iteration if there isn't an existing dispatched iteration for the next task time or
+            // earlier
+            if (s_should_schedule_iteration(
+                    &dispatch_loop->synced_task_data.context->scheduling_state.scheduled_services, next_task_time)) {
+                s_try_schedule_new_iteration(contxt, next_task_time);
+            }
+        }
+    }
+
+    aws_mutex_unlock(&contxt->lock);
+    s_scheduled_service_entry_destroy(entry);
+}
+
+// Iteration function that scheduled and executed by the Dispatch Queue API
+static void s_run_iteration(void *context) {
```

Comment:
The locking here seems a little unnecessarily complex.  If you look at a single call to s_run_iteration:

```
lock(context)
unlock(context)
lock(context) // via begin_iteration, also this seems wrong as you're slurping cross-thread tasks which should be protected by lock(loop) instead
unlock(context) // via begin_iteration
lock(loop)
set thread id/execution
unlock(loop)
run all tasks
lock(loop)
executing = false
unlock(loop)
lock(context) // via end_iteration, again this appears wrong since you access lock(loop) protected data here
unlock(context)
```